gr/lessons/ecto/changesets.md

```diff
@@ -0,0 +1,293 @@
+
+  if name in @fictional_names do
+    changeset
+  else
+    add_error(changeset, :name, "is not a superhero")
+  end
+end
+```
+
+Παραπάνω εισηγάγαμε δύο νέες βοηθητικές συναρτήσεις: τις [`get_field/3`](https://hexdocs.pm/ecto/Ecto.Changeset.html#get_field/3) και [`add_error/4`](https://hexdocs.pm/ecto/Ecto.Changeset.html#add_error/4).
+Η λειτουργία τους είναι αυτονόητη, αλλά σας προτρέπουμε να τους ρίξετε μια ματιά στους συνδέσμους των εγγράφων.
+
+Είναι μια καλή πρακτική να επιστρέφετε πάντα μια δομή `%Ecto.Changeset{}`, ώστε να μπορείτε να χρησιμοποιείτε τον τελεστή `|>` και να κάνετε ευκολότερη την προσθήκη περισσότερων επικυρώσεων αργότερα:
+
+```elixir
+def changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+  |> validate_required([:name])
+  |> validate_length(:name, min: 2)
+  |> validate_fictional_name()
+end
+```
+
+```elixir
+iex> Friends.Person.changeset(%Friends.Person{}, %{"name" => "Bob"})
+%Ecto.Changeset<
+  action: nil,
+  changes: %{name: "Bob"},
+  errors: [name: {"is not a superhero", []}],
+  data: %Friends.Person<>,
+  valid?: false
+>
+```
+
+Θαυμάσια, δουλεύει! Ωστόσο, δεν υπήρχε λόγος να υλοποιήσουμε αυτή τη συνάρτηση μόνοι μας - μπορούσαμε στη θέση της να χρησιμοποιήσουμε τη συνάρτηση `validate_inclusion/4`. Παρόλα αυτά μπορείτε να δείτε πως να προσθέσετε τα δικά σας σφάλματα το οποίο λογικά θα σας φανεί χρήσιμο.
+
+## Προσθέτοντας αλλαγές προγραμματιστικά
+
+Μερικές φορές θα θέλετε να εισάγετε αλλαγές σε ένα σετ αλλαγών χειροκίνητα. Η βοηθητική συνάρτηση `put_change/3` υπάρχει για αυτό το λόγο.
+
+Αντί να κάνουμε το πεδίο `name` απαραίτητο, ας επιτρέψουμε τους χρήστες μας να εγγραφούν χωρίς όνομα, και ας τους ονομάσουμε "Ανώνυμους".
+Η συνάρτηση που χρειαζόμαστε δείχνει οικεία - δέχεται και επιστρέφει ένα σετ αλλαγών, ακριβώς όπως η `validate_fictional_name/1` που δείξαμε νωρίτερα:
+
+```elixir
+def set_name_if_anonymous(changeset) do
+  name = get_field(changeset, :name)
+
+  if is_nil(name) do
+    put_change(changeset, :name, "Anonymous")
+  else
+    changeset
+  end
+end
+```
+
+Μπορούμε να ορίσουμε το όνομα ενός χρήστη σαν "Ανώνυμος" μόνο όταν κάνουν εγγραφή στην εφαρμογή μας. Για να το κάνουν αυτό, θα δημιουργήσουμε μια νέα συνάρτηση δημιουργίας σετ αλλαγών:
+
+```elixir
+def registration_changeset(struct, params) do
+  struct
+  |> cast(params, [:name, :age])
+  |> set_name_if_anonymous()
+end
+```
+
+Τώρα δεν χρειάζεται να περάσουμε ένα `name` και το `Ανώνυμος` θα οριστεί αυτόματα όπως αναμένεται:
+
+```elixir
+iex> Friends.Person.registration_changeset(%Friends.Person{}, %{})
+%Ecto.Changeset<
+  action: nil,
+  changes: %{name: "Anonymous"},
+  errors: [],
+  data: %Friends.Person<>,
+  valid?: true
+>
+```
+
+Η ύπαρξη συναρτήσεων δημιουργίας σετ αλλαγών που έχουν μία συγκεκριμμένη αρμοδιότητα (όπως η `registration_changeset/2`) είναι συνηθισμένη - μερικές φορές χρειαζόμαστε την ελαστικότητα να εκτελούμε συγκεκριμμένες επικυρώσεις ή να φιλτράρουμε συγκεκριμμένες παραμέτρους.
```

Comment:
... συγκεκριμένη αρμοδιότητα ...
... συγκεκριμένες επικυρώσεις ...
... συγκεκριμένες παραμέτρους.

Sorry @gemantzu missed these the first time around :(

Reply:
Thanks @Kosmas !